### PR TITLE
Partially revert connection timeout behavior

### PR DESCRIPTION
**Pull request by nfuller**
Related to issue #1790.

If there is a *connection* timeout the next route should
be tried. Timeouts during a request/response should
probably not be retried.

The timeout/interrupt behavior was changed in PR #1257
for issue #1146. This modified both connection and
request/response behavior. The issue reported was actually
about the use of Thread.interrupt(), not timeouts, but the
behavior change modified both behavior for interrupt and
timeout.

PR #1388 split the handling of exceptions so that separate
code now handles connection errors from that which handles
request/response errors. It faithfully kept the behavior
from PR #1257.

The modification of the timeout behavior during *connection*
caused problems on Android. Now that the two types of error
handling are split it's possible to deal with them
differently and return the connection behavior to trying
the next route. This does not modify behavior during
request/response.

A test has been added for timeouts. It does not seem to be
possible to interrupt a Socket.connect() call with
Thread.interrupt() on either OpenJDK or Android so a
placeholder has been implemented. If it ever is implemented
a Thread.interrupt() during connect will not be retried.

### CODE REVIEW

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 1]
```diff
@@ -164,6 +165,18 @@ public int getPort() {
 
   public String getHostName() {
     maybeStart();
+    if (inetSocketAddress.getAddress() instanceof Inet6Address) {
+      Inet6Address address = (Inet6Address) inetSocketAddress.getAddress();
+      String hostName = inetSocketAddress.getHostName();
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: Does this rally happen?! I think this wants to be a separate PR.

File: okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.SocketChannel;
+
+/**
+ * An {@link Socket} that delegates all calls.
+ */
+public abstract class DelegatingSocket extends Socket {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java]: There isn't a better way to test this? I nolove classes like this, particularly because self-use doesn't work.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java [hunk 3]
```diff
@@ -45,6 +59,14 @@
   // The size of the socket buffers in bytes.
   private static final int SOCKET_BUFFER_SIZE = 256 * 1024;
 
+  // 203.0.113.255 reserved for documentation.
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: That's awesome.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java [hunk 4]
```diff
@@ -45,6 +59,14 @@
...
+  private static final byte[] IPV4_BLACKHOLE_ADDRESS =
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: Shouldn't the type of this be `InetAddress` ?

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java [hunk 5]
```diff
@@ -45,6 +59,14 @@
...
+      new byte[] { (byte) 203, 0, 113, (byte) 255 };
+
+  // 2001:db8::1 reserved for documentation.
+  public static final byte[] IPV6_BLACKHOLE_ADDRESS =
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: Simliarly

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java [hunk 6]
```diff
@@ -70,6 +92,35 @@ protected void configureSocket(Socket socket) throws IOException {
     });
   }
 
+  @Test @Ignore public void interruptConnect() throws Exception {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: Is this true? What about when building an tunnel for HTTPS through an HTTP proxy? That'll fire during InputStream.read().

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java [hunk 7]
```diff
@@ -133,4 +184,103 @@ private void interruptLater(final int delayMillis) {
...
+  }
+
+  private static InetAddress getBlackholeAddress(InetAddress localAddress)
+      throws UnknownHostException {
+    byte[] addressBytes =
+        localAddress instanceof Inet4Address ? IPV4_BLACKHOLE_ADDRESS : IPV6_BLACKHOLE_ADDRESS;
+    return InetAddress.getByAddress(addressBytes);
+  }
+
+  private static class TimeoutCountingSocketFactory extends DelegatingSocketFactory {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: I think you'll want to change DelegatingSocketFactory so that configureSocket returns a Socket. That way we define these 5 overloads only once!

File: okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java (hunk 2 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java]: On further inspection, we're already doing this for lots of things.

Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java]: ... can you make DelegatingSSLSocket extend this?

File: okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.SocketChannel;
+
+/**
+ * An {@link Socket} that delegates all calls.
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java]: nit: An -> A

**Comment by swankjesse** [comment]
LGTM

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java (hunk 1 above)
Comment (nfuller) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: It did happen for me in the test I'd written because it grabs the first non-loopback NIC it finds and then grabs the first local IP address. On my Linux machine it was using IPv6 and everything broke.

i.e.
inetSocketAddress.getHostName() returns "<ipv6 address>%2"

Calling MockWebServer.getUrl() exploded. Looks like scopes are generally not expected. HttpUrl throws IllegalArgumentException and if we got through that URL would throw MalformedURLException unless the % was escaped (and nothing expects the scope later either). Even whether the % should be escaped appears to be a bit controversial.

[I couldn't think of a better way of forcing a connection timeout as loopback addresses can't be "blackholed" with an unroutable address.]

As I said: not sure what to do about it and I agree a second PR once the way forward is clearer.

File: okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java (hunk 2 above)
Comment (nfuller) [inline okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSocket.java]: > ... can you make DelegatingSSLSocket extend this?

Difficult without multiple inheritance because SSLSocket is concrete.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java (hunk 4 above)
Comment (nfuller) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: To switch I think it requires that we handle the UnknownHostException exception from getByAddress(): if we just throw it the class might fail to initialize if we don't have both IPv6 and IPv4. We could set it to null in that case, I suppose.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java (hunk 5 above)
Comment (nfuller) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: As above.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java (hunk 6 above)
Comment (nfuller) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: In this case we know that's not happening because of the way the MockWebServer is configured, but you're right that is not tested.

A lot of the complexity of this change comes about from trying to force blocking behavior so we can then timeout or interrupt it. Not very happy with the test but it seemed better than no test and confirmed the code was broken before and better after.

Forego the test instead?

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java (hunk 7 above)
Comment (nfuller) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: Done for this and others.

**Comment by swankjesse** [comment]
Travis suggests the new test is flaky.
```
Tests run: 4, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 1.195 sec <<< FAILURE! - in com.squareup.okhttp.internal.http.ThreadInterruptTest
timeoutConnect(com.squareup.okhttp.internal.http.ThreadInterruptTest)  Time elapsed: 0.011 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at com.squareup.okhttp.internal.http.ThreadInterruptTest$TimeoutCountingSocketFactory.assertTimeoutCount(ThreadInterruptTest.java:233)
	at com.squareup.okhttp.internal.http.ThreadInterruptTest.timeoutConnect(ThreadInterruptTest.java:122)
```

**Comment by nfuller** [comment]
I am not surprised!

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java (hunk 6 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/ThreadInterruptTest.java]: Yeah.

**Comment by nfuller** [comment]
I have removed the test.
